### ISSUE TITLE
T-SQL: PARTITION BY does not allow expression

### ISSUE THREAD

**Issue by borchero** [sqlfluff/sqlfluff#3938]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

The query below could not be parsed.

### Expected Behaviour

The query should be parseable.

### Observed Behaviour

`sqlfluff lint -d tsql` yields:

```
== [test.sql] FAIL
L:   1 | P:  18 |  PRS | Line 1, Position 18: Found unparsable section: '()
                       | OVER(PARTITION BY CASE WHEN t.col1 = ...'
WARNING: Parsing errors found and dialect is set to 'tsql'. Have you configured your dialect correctly?
All Finished 📜 🎉!
```

### How to reproduce

```sql
SELECT ROW_NUMBER() OVER(PARTITION BY CASE WHEN t.col1 = 'value' THEN 1 END ORDER BY t.col2) rn
FROM mytable t;
```

### Dialect

`tsql`

### Version

1.3.2

### Configuration

n/a

### Are you willing to work on and submit a PR to address the issue?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)
